### ISSUE TITLE
Accessing command option (aka parent option) when there are sub-commands.

### ISSUE THREAD

**Issue by noraj** [davetron5000/gli#330]
The [action instance method](https://www.rubydoc.info/gems/gli/2.21.1/GLI/Command#action-instance_method) offers 3 blocks: `global_options` (root, top level, like `--no-pager` for git), `options` (_last-leaf_ options: that would be the command option if there is just a command, but that would be sub-command options if there is a sub-comment, etc.) and `args` (positional arguments).

Question: how do you access the command options when there is a sub-command?

Given the following example:

```ruby
module ADAssault
  class CLI
    extend GLI::App

    desc 'DNS related commands'
    # since 0.0.1
    command :dns do |dns|
      dns.flag %i[d domain], desc: 'Active Directory domain.', type: String, required: true
      dns.flag %i[s nameserver name-server],
               desc: 'The IP address of the domain DNS server. If not provided uses your system DNS.',
               type: String, required: false
      dns.desc 'Spot all domain controllers in a Microsoft Active Directory environment'
      # since 0.0.1
      dns.command :find_dcs do |find_dcs|
        find_dcs.flag %i[t test], desc: 'test.', type: String, required: false
        find_dcs.action do |_global_options, options, _args|
          dns_opts = options[:nameserver].nil? ? nil : { nameserver: [options[:nameserver]] }
          dcd = ADAssault::DNS::FindDCs.new(options[:domain], dns_opts)
          dcd.display
        end
      end
    end
  end
end
```

The sub-command (find_dcs) help:

```
➜ bundle exec ruby -Ilib -radassault bin/ada dns --domain=THM.local --nameserver=10.10.100.118 find_dcs -h
NAME
    find_dcs - Spot all domain controllers in a Microsoft Active Directory environment

SYNOPSIS

    ada [global options] dns find_dcs [command options] 




COMMAND OPTIONS
    -t, --test=arg - test. (default: none)
```

The command (dns) help:

```
➜ bundle exec ruby -Ilib -radassault bin/ada dns --domain=THM.local --nameserver=10.10.100.118 -h find_dcs
NAME
    dns - DNS related commands

SYNOPSIS

    ada [global options] dns [command options] entry_add

    ada [global options] dns [command options] entry_delete

    ada [global options] dns [command options] entry_update

    ada [global options] dns [command options] find_dcs [-t arg|--test arg]




COMMAND OPTIONS
    -d, --domain=arg                    - Active Directory domain. (required, default: none)
    -s, --nameserver, --name-server=arg - The IP address of the domain DNS server. If not provided uses your system DNS. (default: none)
    -t, --test=arg                      - test. (default: none)




COMMANDS
    entry_add    - TODO
    entry_delete - TODO
    entry_update - TODO
    find_dcs     - Spot all domain controllers in a Microsoft Active Directory environment
```

So I can't access `options[:domain]`, for example, because it is scoped in key which is a parent object.

![image](https://github.com/user-attachments/assets/410d754d-27c8-4167-9d18-8a65c3c6e925)

I would need to access `options[<parent_object>][:domain]` something. After a long moment of debugging and reading https://www.rubydoc.info/gems/gli/2.21.1/GLI/Command, I found that I can call `GLI::Command::PARENT` to access the parent option key, e.g. `options[GLI::Command::PARENT][:domain]`.

I guess GLI was designed to have only one level of commands and that [sub-commands came after](https://github.com/davetron5000/gli/issues/22) in a redesign.

I guess it would be more intuitive to have methods such as `.parent_options` so I could call `find_dcs.parent_options` directly or `.options` so I could call `dns.options`.

Maybe action could have an additional optional block for parent options or even better: arbitrary parent-level options if we could imagine a use for, let's say, 5 levels of sub-commands.

By the way, my `GLI::Command::PARENT` would work if there was one more nesting level, how could I access the parent's parent options? Because I found no method under `find_dcs.parent` to get the parent key ID and there is obviously no constant named `GLI::Command::PARENT_PARENT` nor `GLI::Command::PARENT.parent` method and so on.

So is there a better and easier way to access non-top-level intermediate-level parent options? Else I suggest implementing the previously mentioned ideas (methods and/or block).

**Comment by noraj** [comment]
My temporary workaround:

```ruby
# frozen_string_literal: true

require 'adassault/dns'
require 'gli'

module ADAssault
  class CLI
    extend GLI::App

    desc 'DNS related commands'
    # since 0.0.1
    command :dns do |dns|
      dns.flag %i[d domain], desc: 'Active Directory domain.', type: String, required: true
      dns.flag %i[s nameserver name-server],
               desc: 'The IP address of the domain DNS server. If not provided uses your system DNS.',
               type: String, required: false
      dns.desc 'Spot all domain controllers in a Microsoft Active Directory environment'
      # since 0.0.1
      dns.command :find_dcs do |find_dcs|
        find_dcs.action do |_global_options, options, _args|
          parent_options = options[GLI::Command::PARENT]
          dns_opts = parent_options[:nameserver].nil? ? nil : { nameserver: [parent_options[:nameserver]] }
          dcd = ADAssault::DNS::FindDCs.new(parent_options[:domain], dns_opts)
          dcd.display
        end
      end
    end
  end
end
```

**Comment by davetron5000** [comment]
Originally, GLI tried to merge everything into one hash but it was way too complicated.  The `PARENT` thing is what I could come up with.  In theory if you had several levels, `PARENT` would be at each one:

https://github.com/davetron5000/gli/blob/19eae87ca23ab03237b3533833aa51170ea30e84/test/apps/todo/lib/todo/commands/make.rb#L30-L35

Essentially, your app needs to know where each flag is configured.  I'll be honest, more than 2 or 3 levels of nesting and things get really confusing.

Another option is to put all your non-global flags & switches at the "leaf" commands.  This may require duplication, but you can put that into a method, e.g. 

```ruby
def add_leaf_command_flags(leaf_command)
  leaf_command.flag :foo 
  leaf_command.switch :bar
end
```

**Comment by noraj** [comment]
> I'll be honest, more than 2 or 3 levels of nesting and things get really confusing.

Yeah more than 3 levels may be exaggerated. But I have use of 3 levels for the tool I'm building. For two level tools, they could often have a use for 3 levels instead of having tons of sub-commands like `user_add`, `user_delete`, `user_delete` etc. (`tool unix user add` instead of `tool unix user_add`). The helper I was talking about could be useful for 3 levels.

As 3 levels doesn't sound unreasonable, is it a won't fix for the helpers I suggested? (`.parent_options` or `parent.options` and `.options`)
I could even work on it and MR if you are ok to merge this?

**Comment by davetron5000** [comment]
It feels somewhat special-case to me, plus most would be using hashes so methods won't work in those cases anyway.  Your use-case is the first one I've seen that has the middle command having its own options (even though GLI has always supported it). I built a large CLI app with 3 levels and we ended up having flags only global or on the leaves, so I'm not yet keen to make a big change for one use case.
